lib/ModuleFinder.js

```
@@ -0,0 +1,223 @@
+      .then(() => this.storage.purgeDeadNodeModules(this.workingDirectory));
+  }
+
+  startWatcher() {
+    return this.watcher.initialize();
+  }
+
+  processQueue(done) {
+    const file = this.queue.pop();
+    if (!file) {
+      this.processingQueue = false;
+      logLevel.debug('Queue empty');
+      done();
+      return;
+    }
+    const { path: pathToFile, mtime, alias } = file;
+    this.processingQueue = true;
+    logLevel.debug(`Processing ${pathToFile}`);
+    const fullPath = path.join(this.workingDirectory, pathToFile);
+    readFile(fullPath).then((data) => {
+      let exports = { named: [], hasDefault: true };
+      try {
+        exports = findExports(data, fullPath);
+      } catch (e) {
+        logLevel.error(`Failed to parse ${pathToFile}`, e.message, e.stack);
+      }
+      if (exports.named.length || exports.hasDefault) {
+        const all = exports.named.slice(0);
+        if (exports.hasDefault) {
+          all.push('default');
+        }
+        logLevel.debug(
+          `Found exports for ${pathToFile}:\n${all.join(', ')}`);
+      } else {
+        logLevel.debug(`No exports found for ${pathToFile}`);
+      }
+      const defaultNames = [];
+      if (exports.hasDefault) {
+        if (alias) {
+          defaultNames.push(...aliasedExportNames(alias, this.ignorePackagePrefixes));
+        } else {
+          defaultNames.push(...defaultExportNames(pathToFile));
+        }
+      }
+      this.storage.update({
+        names: exports.named,
+        defaultNames,
+        pathToFile,
+        mtime,
+      }).then(() => {
+        this.processQueue(done);
+      }).catch((error) => {
+        logLevel.error(
+          `Failed to update ${pathToFile} in storage. Reason: ${error.message}`);
+        this.processQueue(done);
+      });
+    }).catch((err) => {
+      logLevel.error(err.message);
+      this.processQueue(done);
+    });
+  }
+
+  handleFilesAdded(unexpandedFiles) {
+    return new Promise((resolve) => {
```

With all of these promises, it might be worth spending some time going through and making sure that errors aren't being swallowed.